### PR TITLE
3849 win conditions fixes

### LINKED ISSUE #2121: Sink ships when a player has been defeated
When a player has lost all warehouses, there can still be expeditions ships sailing around the map forever.

So, after a a good while that would allow to colonize any remaining open port spaces (timeout to be determined, or maybe a port space check?), the ships should be sunk.

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2018-10-13T09:03:05Z
 * owner: gunchleoc
 * the launchpad url was https://bugs.launchpad.net/bugs/1797682

### LINKED ISSUE #3190: "You are defeated" - message is displayed too early
The "You are defeated" message is also displayed if the player has a warehouse constructionsite which has all his wares delivered, or if he has an expedition. So, it is possible to get the FoW removed and to see what the other tribes are doing while having a functional infrastructure.

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2017-02-26T12:19:09Z
 * owner: fulgen
 * the launchpad url was https://bugs.launchpad.net/bugs/1668047

### LINKED ISSUE #3849: Player can be resurrected and has no fog of war.
**Describe the bug**
We had a very interesting bug. A player got defeated because his last warehouse got destroyed and then relegated to spectator. The fog of war was lifted, and he could see everything. He had a few minutes after he got defeated a warehouse that was under construction was finished and the player got back into the match.

This was kind of unexpected. Also, the resurrected player had no fog of war anymore and had an advantage over the others.

Edit:

The player also can look into others buildings but cannot attack.

### PR DESCRIPTION
**Pull request by hessenfarmer**
fixes #3849 
fixes #3190
fixes #2121
reduces the effect of #1037

supersedes #3742

### CODE REVIEW

File: src/scripting/lua_game.cc [hunk 1]
```diff
@@ -902,6 +903,68 @@ int LuaPlayer::get_buildings(lua_State* L) {
...
+      case, the method returns an array of all Constructionssites that the player has
+      of this kind. If which is an array, the function returns a table of
+      (name,array of buildings) pairs.
+
+      :type which: name of constructionsites building or array of building names
+      :rtype which: :class:`string` or :class:`array`
+      :returns: information about the players constructionsites
+      :rtype: :class:`array` or :class:`table`
+*/
+int LuaPlayer::get_constructionsites(lua_State* L) {
```
Comment (Noordfrees) [inline src/scripting/lua_game.cc:878]: This function is 99% identical to `LuaPlayer::get_buildings`. Please pull the common code out into a new helper function `LuaPlayer::do_get_buildings(lua_State* L, const bool csites)` and change the two Lua hooks into one-liners that invoke the helper.

File: src/scripting/lua_game.cc [hunk 2]
```diff
@@ -902,6 +903,68 @@ int LuaPlayer::get_buildings(lua_State* L) {
 	return 1;
 }
 
+/* RST
+   .. method:: get_constructionsites(which)
+
+      which can be either a single name or an array of names. In the first
+      case, the method returns an array of all Constructionssites that the player has
```
Comment (Noordfrees) [inline src/scripting/lua_game.cc]:
```suggestion
      case, the method returns an array of all constructionsites that the player has
```

**Review by Noordfrees — COMMENTED** [review]
Can you explain what the difference between *Endless Game* and *No Rules* is? Do we really need both?

File: data/scripting/win_conditions/hq_hunter.lua [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+push_textdomain("win_conditions")
+
+include "scripting/win_conditions/win_condition_texts.lua"
+
+local wc_name = "HQ Hunter"
+-- This needs to be exactly like wc_name, but localized, because wc_name
+-- will be used as the key to fetch the translation in C++
+local wc_descname = _("HQ Hunter")
+local wc_version = 2
+local wc_desc = _ "The tribe or team that can destroy all other headquarters wins the game!. This wincondition does not work with the following starting conditions: Poor Hamlet, Fortified Village, Discovery and New World."
```
Comment (klaus-halfmann) [inline data/scripting/win_conditions/hq_hunter.lua]: Better just "works only when you start with Headqueaers", will have to be adopted with every other new win condition.
Could the script start with chehcking if every player has a HQ and send a message if not?

**Review by klaus-halfmann — COMMENTED** [review]
Just a singele inline comment

File: data/scripting/win_conditions/hq_hunter.lua (hunk 3 above)
Comment (hessenfarmer) [reply data/scripting/win_conditions/hq_hunter.lua]: well i saw this note as a workaround until we mange to check dependencies of Startconditons with Winconditions in c++.
So I found it better to tell exactly what does not work, cause e.g village or trading outpost or even struggling outpost will work which isn't obvious to the player from their names

**Comment by hessenfarmer** [comment]
> 
> 
> Can you explain what the difference between _Endless Game_ and _No Rules_ is? Do we really need both?

The differnce is in Endless game the rule when a player is defeated is in place (i.e. the script will destroy the Milsites and constructionsites and ships of a defeated player and give him full sight) while really no rule is in place for "no Rules" allowing full resurrection in the discussed cornercases

**Comment by Noordfrees** [comment]
Well, if *Endless Game* now marks players as thoroughly defeated there is no significant difference to *Autocrat* except that the only player left at the end won't get a congrats message. I'd suggest to remove the new *No Rules* again and change both *Endless Game* and *Endless Game Fogless* to the no-rules behaviour as I can see no value gained by offering such fine degrees of distinction to the players.

**Comment by hessenfarmer** [comment]
> 
> 
> Well, if _Endless Game_ now marks players as thoroughly defeated there is no significant difference to _Autocrat_ except that the only player left at the end won't get a congrats message. I'd suggest to remove the new _No Rules_ again and change both _Endless Game_ and _Endless Game Fogless_ to the no-rules behaviour as I can see no value gained by offering such fine degrees of distinction to the players.

ok Will do so.

File: src/scripting/lua_game.cc (hunk 1 above)
Comment (hessenfarmer) [reply src/scripting/lua_game.cc:878]: done

File: src/scripting/lua_game.cc [hunk 4]
```diff
@@ -1125,6 +1098,70 @@ int LuaPlayer::allow_forbid_buildings(lua_State* L, bool allow) {
...
+      of this kind (returns buildings or constructionsites dependent on csites). If which
+      is an array, the function returns a table of (name,array of buildings) pairs.
+
+      :type which: name of constructionsites building or array of building names
+      :rtype which: :class:`string` or :class:`array`
+      :type csites: whether to return constructionsites
+      :rtype which: :class:`boolean`
+      :returns: information about the players constructionsites
+      :rtype: :class:`array` or :class:`table`
+*/
```
Comment (Noordfrees) [inline src/scripting/lua_game.cc]:
```suggestion
```
This function is not exposed to Lua scripters so it must not have an RST doc

File: data/scripting/win_conditions/no_rules.lua [hunk 5]
```diff
@@ -0,0 +1,28 @@
+-- =======================================================================
```
Comment (Noordfrees) [inline data/scripting/win_conditions/no_rules.lua]: this file can also be deleted now

Comment (hessenfarmer) [reply data/scripting/win_conditions/no_rules.lua]: done

**Comment by frankystone** [comment]
When starting an endless game i get this:
~~~~
[/home/kaputtnik/Quellcode/widelands-repo/widelands/src/scripting/lua_errors.cc:22] [string "scripting/win_conditions/endless_game.lua"]:24: attempt to call a nil value (global 'broadcast_objective')
~~~~

**Comment by hessenfarmer** [comment]
> 
> 
> When starting an endless game i get this:
> 
> ```
> [/home/kaputtnik/Quellcode/widelands-repo/widelands/src/scripting/lua_errors.cc:22] [string "scripting/win_conditions/endless_game.lua"]:24: attempt to call a nil value (global 'broadcast_objective')
> ```

Oops, seems like I deleted to much includes. Will be fixed tonight. Thanks for finding this

**Comment by hessenfarmer** [comment]
I tried to implement a crosscheck in the winconditions dropdown but I failed to identify how to determine the actually chosen startcondition. Any ideas?

**Comment by Noordfrees** [comment]
Do you want to implement that selecting the HQ hunter win condition disables the possibility to select certain starting conditions, or that selecting a certain starting condition disables the possibility to select the HQ hunter win condition? (Note that each player has an own starting condition so several may be selected.)
We already have code to selectively disable win conditions (artifacts) and starting conditions (discovery, new world) based on the selected map's tags, and to en-/disable peaceful mode depending on the selected win condition. This could be extended for this purpose.

**Comment by hessenfarmer** [comment]
Ideally this should be either way round depending on the choice that was made first. I saw the code for the map tags in both ways however I didn't see where the choice of the player was stored

**Comment by Noordfrees** [comment]
Best access it indirectly via `settings_.settings().players[i].initialization_index`
see also `logic/game_settings.h` and `logic/map_objects/tribes/tribe_basic_info.h`

**Comment by frankystone** [comment]
Played a 12 hrs game with HQ-Hunter:
hq_hunter_won

Anything  more to test?

**Comment by hessenfarmer** [comment]
thanks for the test. I'll give it a try to implement c++ stuff but if I fail I think everyzthing has been tested so far. 
Were militarysites and constructionsites burned correctly?

File: data/scripting/win_conditions/hq_hunter.lua [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                        p:get_buildings("atlanteans_warehouse"),
+                        p:get_buildings("empire_warehouse"),
+                        p:get_buildings("frisians_warehouse"),
+                        p:get_buildings("amazons_warehouse"),
+                        p:get_buildings("barbarians_port"),
+                        p:get_buildings("atlanteans_port"),
+                        p:get_buildings("empire_port"),
+                        p:get_buildings("frisians_port"),
+                        p:get_buildings("amazons_port")
+                    )) do
```
Comment (frankystone) [inline data/scripting/win_conditions/hq_hunter.lua]: Can the hardcoded values be prevented? Something like:

~~~~
warehouses_and_ports = {}
for i, pl in ipairs(wl.Game().players) do
   for j, building in ipairs(pl.tribe.buildings) do
       if building.type_name == "warehouse" then
           table.insert(warehouses_and_ports, building)
~~~~
Then feed `p:get_buildings(warehouses_and_ports)`
This do not work for the names of  headquarters though.

**Comment by frankystone** [comment]
> thanks for the test. I'll give it a try to implement c++ stuff but if I fail I think everyzthing has been tested so far.
> Were militarysites and constructionsites burned correctly?

Good question :-)

But i think yes. It was irritating that buildings under half seen areas do not change. So one can see a darkened port but if a ship goes by and one sees the reality the port was no there anymore.

File: data/scripting/win_conditions/hq_hunter.lua (hunk 6 above)
Comment (hessenfarmer) [reply data/scripting/win_conditions/hq_hunter.lua]: excellent Idea. will do this

**Comment by hessenfarmer** [comment]
although the code in winconditions_functions.lua works it is not failsafe as it produces a coroutine error if a building to be destroyed is not present anymore which can happen due to different miltitarysites influence. I need to think about how to get this safe.

**Comment by frankystone** [comment]
This bothers me as well: The only way i know to check if an immovable (like a building) exists anymore is done via 
~~~~
if field.immovable then 
     do stuff with immovable
end
~~~~

Those checks blows up the code. Especially because it is only related to a `field`. It would be nice to have such a check internally. Say if one calls `immovable.destroy()` the error should be hidden (not thrown) if this building (or other immovable) doesn't exist anymore.

**Comment by Noordfrees** [comment]
This would cause weird bugs as scripters could expect some kind of error when calling a method on a non-existing object. But we could implement a property `exists` that just checks whether a Lua variable still refers to a valid map object, and provide a short Lua helper function:
```
function destroy_if_exists(immovable)
   if immovable and immovable.exists then immovable:destroy() end
end
```

**Comment by frankystone** [comment]
Now it not possible anymore to build a warehouse. Close after it get finished it will be magically destroyed.

[Savegame close before the Warehouse is  built](https://github.com/widelands/widelands/files/5925392/test_hq_hunter.zip)

**Comment by hessenfarmer** [comment]
> Now it not possible anymore to build a warehouse. Close after it get finished it will be magically destroyed.
> 
> [Savegame close before the Warehouse is built](https://github.com/widelands/widelands/files/5925392/test_hq_hunter.zip)

thanks for testing should be fixed now

**Comment by frankystone** [comment]
Got this after conquering the last headquarters:

~~~~
[02:38:36.400 game] ERROR: Error in Lua Coroutine
[02:38:36.400 game] ERROR: [/home/kaputtnik/Quellcode/widelands-repo/widelands/src/scripting/lua_errors.cc:22] MapObject no longer exists!
[02:38:36.400 game] ERROR: stack traceback:
[02:38:36.400 game] ERROR:      [string "scripting/win_conditions/win_condition_functi..."]:98: in global 'check_player_defeated'
[02:38:36.400 game] ERROR:      [string "scripting/win_conditions/hq_hunter.lua"]:33: in function <[string "scripting/win_conditions/hq_hunter.lua"]:23>
~~~~

My example was not good. Instead of the building, just store the field:

~~~~
warehouses_and_ports = {}
for i, pl in ipairs(wl.Game().players) do
   for j, building in ipairs(pl.tribe.buildings) do
       if building.type_name == "warehouse" then
           table.insert(warehouses_and_ports, building.fields[1]) -- the important change
~~~~

Instead of the immovable store the field where this immovable stands on. Then you could do something like this:

~~~~
if field.immovable then
   field.immovable:destroy()
~~~~

Does this help?

**Comment by hessenfarmer** [comment]
> 
> 
> Got this after conquering the last headquarters:
> 
> ```
> [02:38:36.400 game] ERROR: Error in Lua Coroutine
> [02:38:36.400 game] ERROR: [/home/kaputtnik/Quellcode/widelands-repo/widelands/src/scripting/lua_errors.cc:22] MapObject no longer exists!
> [02:38:36.400 game] ERROR: stack traceback:
> [02:38:36.400 game] ERROR:      [string "scripting/win_conditions/win_condition_functi..."]:98: in global 'check_player_defeated'
> [02:38:36.400 game] ERROR:      [string "scripting/win_conditions/hq_hunter.lua"]:33: in function <[string "scripting/win_conditions/hq_hunter.lua"]:23>
> ```

A replay of this would be nice. Note that this did not happen within hq_hunter.lua but in the winconditions_functions.lua and is exactly the instability I described above.

> My example was not good. Instead of the building, just store the field:
> 
> ```
> warehouses_and_ports = {}
> for i, pl in ipairs(wl.Game().players) do
>    for j, building in ipairs(pl.tribe.buildings) do
>        if building.type_name == "warehouse" then
>            table.insert(warehouses_and_ports, building.fields[1]) -- the important change
> ```
> 
> Instead of the immovable store the field where this immovable stands on. Then you could do something like this:
> 
> ```
> if field.immovable then
>    field.immovable:destroy()
> ```
> 
> Does this help?

Well, the idea is not bad. however p.tribe. buildings is a list of building descriptions and not the actual building. So I need to identifiy the field of an existing building. Will try it out.

**Comment by frankystone** [comment]
> Well, the idea is not bad. however p.tribe. buildings is a list of building descriptions and not the actual building.

Oh, yes, my bad.

**Comment by hessenfarmer** [comment]
> 
> 
> > Well, the idea is not bad. however p.tribe. buildings is a list of building descriptions and not the actual building.
> 
> Oh, yes, my bad.

No worries. You brought me to the right solution. Many thanks for suggesting.  
This should be safeguarded from the issue now.  

I am not sure whether I find time to implement the c++ stuff but in the current state it will safely loose the game in fortified village after 5 sec. Discovery will be lost after the first port is built.   

Edit: and the resulting code is even more streamlined and I didn't need the sleep in the beginning anymore

**Comment by frankystone** [comment]
Maybe not related to this branch but can we have another text for autocrat?

Currently:
> The tribe or team that can defeat all others wins the game!

Suggestion:
> The tribe or team which destroys all warehouses and headquarters of opponents wins the game!

**Comment by hessenfarmer** [comment]
> 
> 
> Maybe not related to this branch but can we have another text for autocrat?
> 
> Currently:
> 
> > The tribe or team that can defeat all others wins the game!
> 
> Suggestion:
> 
> > The tribe or team which destroys all warehouses and headquarters of opponents wins the game!

in this case we would need to add ports as well, which make it even longer to read. I would stay with defeat as from my perspective we don't need to explain defeated in the helptext for the wincondition. People tend to not read text anymore ;-)

**Comment by frankystone** [comment]
> in this case we would need to add ports as well, which make it even longer to read. I would stay with defeat as from my perspective we don't need to explain defeated in the helptext for the wincondition. People tend to not read text anymore ;-)

Yes, of course. But clarifying the win condition would help a lot, at least for new players. I remember that i didn't know the exactly definition when starting playing widelands and thought i had to destroy all military buildings. If it is written some where, we can always ask: "Have you read that?" :-D

> The tribe or team which destroys all warehouses, ports and headquarters of opponents wins the game!

Isn't as long as:

> "The tribe or team that can destroy all other headquarters wins the game!. This wincondition does not work with the following starting conditions: Poor Hamlet, Fortified Village, Discovery and New World."

**Comment by hessenfarmer** [comment]
@Noordfrees 
the last coomit does not work, cause I am not able to determine how to access the currently selected wincondition.
2 Questions?
1. has my concept any chance to work?
2. how can I access the current selected wc from singleplayerdropdown (and the equivalent for MP as well)

**Comment by Noordfrees** [comment]
I did not look at it closely yet but I guess
1. yes
2. Did you try `settings_->settings().win_condition_script`? 
    I believe this should work at least for MP; for SP you might need to add `settings_.set_win_condition_script(win_condition_dropdown_.get_selected());` 
   to `LaunchSPG::win_condition_selected()` to get it working there as well.

**Comment by hessenfarmer** [comment]
> 
> 
> I did not look at it closely yet but I guess
> 
>     1. yes
> 
>     2. Did you try `settings_->settings().win_condition_script`?
>        I believe this should work at least for MP; for SP you might need to add `settings_.set_win_condition_script(win_condition_dropdown_.get_selected());`
>        to `LaunchSPG::win_condition_selected()` to get it working there as well.

in my first attempt I tried `if (wc == settings_->get_win_condition_script())` but this is not set yet while the dropdown is still displayed. So I need to use the currently selected entry in the dropdownlist.  
Unfortunately my c++ abilities are very limited and I mostly do copy and paste of code blocks which I think I understand.  
In this case it seems I am lost so I wouldn't object if you would be so kind to fix my attempt

**Comment by Noordfrees** [comment]
Okay I'll have a try later

**Comment by hessenfarmer** [comment]
> 
>     2. Did you try `settings_->settings().win_condition_script`?

Update this does not work as I believe it is not accessing the dropdown but the final result as well

**Comment by Noordfrees** [comment]
Well. I had a look at this with GDB and it seems there's a name mismatch. The `incompatible_wc` is a name ("HQ Hunter") but the `win_condition_script` is a filepath ("scripting/win_conditions/hq_hunter.lua"). Changing this in the Lua files seems to fix it on a quick test.

**Comment by hessenfarmer** [comment]
I retried `if (wc == settings_->get_win_condition_script())` and it works in multiplayer as well as `settings_->settings().win_condition_script`.

However I believe we need to call an update of the start conditions if the wc is changed. Curently after a change we need to change the tribe for this to take effect.

**Comment by Noordfrees** [comment]
For SP, just call `player_setup_.update()`. You may need to move `update()` to the `public` section in `SinglePlayerSetupBox`.
For MP, something similar with `mpsg_.update_players()` and `MultiPlayerSetupGroup::update_players()` I guess.

**Comment by hessenfarmer** [comment]
ok it is working in Multiplayer now.
just singleplayer left

**Comment by hessenfarmer** [comment]
ready for review now. Thanks for the hints now it works fine from my perspective.

If no othger comments I'll do the documentation in the startconditions part and remove the long text from HQ Hunter.

File: data/scripting/win_conditions/hq_hunter.lua [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+push_textdomain("win_conditions")
+
+include "scripting/win_conditions/win_condition_texts.lua"
+
+local wc_name = "HQ Hunter"
+-- This needs to be exactly like wc_name, but localized, because wc_name
+-- will be used as the key to fetch the translation in C++
+local wc_descname = _("HQ Hunter")
+local wc_version = 2
+local wc_desc = _ "The tribe or team that can find and destroy all other headquarters wins the game!. Does not work with starting conditions without headquarters."
```
Comment (Noordfrees) [inline data/scripting/win_conditions/hq_hunter.lua]:
```suggestion
local wc_desc = _ "The tribe or team that can find and destroy all other headquarters wins the game! Does not work with starting conditions without headquarters."
```

File: src/ui_fsmenu/singleplayerdropdown.cc [hunk 8]
```diff
@@ -310,7 +311,14 @@ void SinglePlayerStartTypeDropdown::fill() {
...
 		}
-		if (matches_tags) {
+		bool compatible_wc = true;
+		for (const std::string& wc : addme.incompatible_win_conditions) {
+			if (wc == settings_->settings().win_condition_script) {
+				compatible_wc = false;
+				break;
+			}
+		}
+		if (matches_tags && compatible_wc) {
```
Comment (Noordfrees) [inline src/ui_fsmenu/singleplayerdropdown.cc]:
```suggestion
		if (matches_tags && !addme.incompatible_win_conditions.count(settings_->settings().win_condition_script)) {
```

File: src/logic/map_objects/tribes/tribe_basic_info.cc [hunk 9]
```diff
@@ -66,15 +74,22 @@ TribeBasicInfo::TribeBasicInfo(std::unique_ptr<LuaTable> table)
...
 				if (script_table->has_key("map_tags")) {
 					std::unique_ptr<LuaTable> t = script_table->get_table("map_tags");
 					for (int key : t->keys<int>()) {
 						tags.insert(t->get_string(key));
 					}
 				}
-				initializations.push_back(Initialization(script_path,
-				                                         script_table->get_string("descname"),
-				                                         script_table->get_string("tooltip"), tags));
+				if (script_table->has_key("incompatible_wc")) {
```
Comment (Noordfrees) [inline src/logic/map_objects/tribes/tribe_basic_info.cc:88]: That's the same code as above, and as in `gameclient`. Please add a TODO that these three Initialization creation pathways should be pulled out into a common function.

File: src/ui_fsmenu/multiplayersetupgroup.cc [hunk 10]
```diff
@@ -468,7 +468,14 @@ struct MultiPlayerPlayerGroup : public UI::Box {
...
 				}
-				if (matches_tags) {
+				bool compatible_wc = true;
+				for (const std::string& wc : addme.incompatible_win_conditions) {
+					if (wc == settings_->get_win_condition_script()) {
+						compatible_wc = false;
+						break;
+					}
+				}
+				if (matches_tags && compatible_wc) {
```
Comment (Noordfrees) [inline src/ui_fsmenu/multiplayersetupgroup.cc]:
```suggestion
				if (matches_tags && !addme.incompatible_win_conditions.count(settings_->get_win_condition_script())) {
```

**Review by Noordfrees — COMMENTED** [review]
Two more comments, then this is good to go codewise. Did not test it much though so @frankystone could you approve the PR please.

File: src/ui_fsmenu/multiplayersetupgroup.cc (hunk 10 above)
Comment (hessenfarmer) [reply src/ui_fsmenu/multiplayersetupgroup.cc]: @Nordfrees 
thanks for the suggestion. this is the same code as in singleplayer with the only difference of there I used settings_->settings().win_condition_script and here I tried settings_->get_win_condition_script(). 
While they are delivering the same result is there any advantage/disadvantage in the syntax as I think we should harmonize this.

**Comment by hessenfarmer** [comment]
One last question we do not have a documentation of the startcondition scripts yet. I would try to insert one in atlanteans folder. 
I think I need to add the RST section to the script and the path of the script to our list of doc files. Correct?

File: src/ui_fsmenu/multiplayersetupgroup.cc (hunk 10 above)
Comment (Noordfrees) [reply src/ui_fsmenu/multiplayersetupgroup.cc]: The function simply returns the member variable so it doesn't really matter, but a high-level function (`settings_->get_win_condition_script()`) is usually better style than directly accessing low-level variables.

**Review by frankystone — APPROVED** [review]
Testing is fine here.

**Pull request merged**